content/pda.md

```diff
@@ -0,0 +1,523 @@
+
+```rust
+ pub fn find_program_address(seeds: &[&[u8]], program_id: &Pubkey) -> (Pubkey, u8) {
+        Self::try_find_program_address(seeds, program_id)
+            .unwrap_or_else(|| panic!("Unable to find a viable program address bump seed"))
+    }
+```
+
+The `try_find_program_address` function then introduces the `bump_seed`. The `bump_seed` is a `u8` variable with a value ranging between 0 to 255. In a loop from 255 to 0, a `bump_seed` is appended to the optional input seeds and then passed to the `create_program_address` function. If the output of `create_program_address` is not a valid PDA, then the `bump_seed` is decreased by 1 and the loop continues.
+
+```rust
+pub fn try_find_program_address(seeds: &[&[u8]], program_id: &Pubkey) -> Option<(Pubkey, u8)> {
+
+    let mut bump_seed = [std::u8::MAX];
+    for _ in 0..std::u8::MAX {
+        {
+            let mut seeds_with_bump = seeds.to_vec();
+            seeds_with_bump.push(&bump_seed);
+            match Self::create_program_address(&seeds_with_bump, program_id) {
+                Ok(address) => return Some((address, bump_seed[0])),
+                Err(PubkeyError::InvalidSeeds) => (),
+                _ => break,
+            }
+        }
+        bump_seed[0] -= 1;
+    }
+    None
+
+}
+```
+
+The `create_program_address` function performs a set of hash operations over the seeds and `program_id` to compute a key and then verifies if the computed key lies on the ed25519 elliptic curve or not. If a valid PDA is found, then the PDA is returned. Otherwise, an error is returned.
+
+```rust
+pub fn create_program_address(
+        seeds: &[&[u8]],
+        program_id: &Pubkey,
+    ) -> Result<Pubkey, PubkeyError> {
+
+        let mut hasher = crate::hash::Hasher::default();
+        for seed in seeds.iter() {
+            hasher.hash(seed);
+        }
+        hasher.hashv(&[program_id.as_ref(), PDA_MARKER]);
+        let hash = hasher.result();
+
+        if bytes_are_curve_point(hash) {
+            return Err(PubkeyError::InvalidSeeds);
+        }
+
+        Ok(Pubkey::new(hash.as_ref()))
+
+    }
+```
+
+In summary, the `find_program_address` function passes our input seeds and `program_id` to the `try_find_program_address` function. The `try_find_program_address` function adds a `bump_seed` (starting from 255) to our input seeds calls the `create_program_address` function in a loop until a valid PDA is found. Once a valid PDA is found, both the PDA and `bump_seed` are returned.
+
+Note that it is possible for the same input seeds and different valid bumps to generate different valid PDAs. The `bump_seed` returned by `find_program_address` will always be the first valid PDA found, and thus the `bump_seed` with the largest value. This `bump_seed` is commonly referred to as the "canonical bump". It's recommended to only use the canonical bump to avoid confusion and alway validate the PDAs passed into your program.
```

Comment:
@ZYJLiu The opening sentence is a little awk, sugg:
> Note that for the same input seeds, different valid bumps will generate different valid PDAs.

The second sentence is a bit runon and the final is also awk, sugg:
> The `bump_seed` returned by `find_program_address` will always be the first valid PDA found.  Because the function starts with a `bump_seed` value of 255 and iterates downwards to zero, the `bump_seed` that ultimately gets returned will always be the largest valid 8-bit value possible. This `bump_seed`... "canonical bump". To avoid confusion, it's recommended to only use the canonical bump, and to always validate every PDA passed into your program.